### ISSUE TITLE
Scheduler slows down at high number of now scheduled tasks

### ISSUE THREAD

**Issue by fischi1401** [kagkarlsson/db-scheduler#127]
Having a huge number of tasks scheduled for asynchronous/parallel calculations results in a significant application slowdown.

The select statements in JdbcTaskRepository don't have a limit, so all scheduled tasks are loaded when they are due for execution.

Maybe it would make sense to set a query limit with the number of scheduler threads or what the number of application instances in a cluster would need? Or maybe as optional scheduler setting?

I temporarily solved my performance my issue by manually updating the execution time and setting it to a evenly distributed timerange: 
`UPDATE scheduled_tasks SET execution_time = NOW() + (random() * (NOW()+'+4 days' - NOW()));` and added an index on `execution_time` and `task_name, task_instance, version`
Using a queue would have been the better solution for me but more implementation effort.

**Comment by kagkarlsson** [comment]
I am a bit surprised that you have this issue. I have tested with millions of executions due 🤔  (I did have to create an index for `execution_time` though)

There is a [limit](https://github.com/kagkarlsson/db-scheduler/blob/master/db-scheduler/src/main/java/com/github/kagkarlsson/scheduler/JdbcTaskRepository.java#L137) on how many executions are fetched, it is by default 3 * <nr-of-threads> I think.

**Comment by kagkarlsson** [comment]
Could it be that you just need the index?

**Comment by kagkarlsson** [comment]
Currently, db-scheduler expects that the thread executes the execution fully and not just adds it to some async executor-pool

**Comment by fischi1401** [comment]
You are right. 
The issue was on my side. Probably I overloaded the database that it got slow. 
Then the pg_stat_activity was showing a long execution time for the select and I wasn't seeing a limit in the query there: `select * from scheduled_tasks where picked = $1 and execution_time <= $2  order by execution_time asc` so I draw the conclusion that it's missing. I didn't see it in the code at a quick glance as well but you do it on the prepared statement after the SQL query template.
I changed the execution time to now for my tasks and it's working fine with a million of tasks.
So it's all good. No issue. 
Great project by the way! Thank you very much for that!
